### PR TITLE
Insert Fixes

### PR DESCRIPTION
**Pull request by Hydrocharged**
Work so far for fixing issues in imports.

### CODE REVIEW

**Comment by Hydrocharged** [comment]
This doesn't include anything dealing with existing primary keys nor insert ignore.

**Comment by Hydrocharged** [comment]
This felt a bit hacky to do. I looked into how SELECT statements work, thinking I could make use of some of that logic to do the entire check on this side. However, I see it uses an iterator, which is constructed from a parsed statement, and I've no idea how to extract what I need from InsertInto to get to that state.

Instead, I decided to put the check in the Insert function itself. I toyed around with making a separate function at first (one that was called before actually calling Insert), but found that there would be duplicate work at least in our case, and attempting to cache that would add more complexity than necessary. Also not sure if I should use a boolean, or perhaps a special error that is handled differently.

File: sql/plan/insert.go [hunk 1]
```diff
@@ -67,6 +75,19 @@ func (p *InsertInto) Execute(ctx *sql.Context) (int, error) {
 
 	dstSchema := p.Left.Schema()
 	projExprs := make([]sql.Expression, len(dstSchema))
+
+	if len(p.Columns) == 0 {
```
Comment (zachmu) [inline sql/plan/insert.go]: Comment here about the no column case

File: sql/plan/insert.go [hunk 2]
```diff
@@ -129,7 +160,19 @@ func (p *InsertInto) WithChildren(children ...sql.Node) (sql.Node, error) {
...
+func (p *InsertInto) ValidateValueCount(ctx *sql.Context) error {
+	switch node := p.Right.(type) {
+	case *Values:
+		for _, exprTuple := range node.ExpressionTuples {
+			if len(exprTuple) != len(p.Columns) {
+				return ErrInsertIntoMismatchValueCount.New()
+			}
+		}
+	}
+	return nil
```
Comment (zachmu) [inline sql/plan/insert.go]: Better to return an error here than let an unsupported p.Right give bogus results.

File: sql/plan/insert.go [hunk 3]
```diff
@@ -102,10 +126,17 @@ func (p *InsertInto) Execute(ctx *sql.Context) (int, error) {
 			return i, err
 		}
 
-		if err := insertable.Insert(ctx, row); err != nil {
+		rowDidNotExist, err := insertable.Insert(ctx, row)
```
Comment (zachmu) [inline sql/plan/insert.go]: As discussed above, you should handle the ignore keyword by just ignoring any error here (and not closing the iterator). Don't try to get into the semantics of duplicate keys yet.

File: sql/plan/insert.go [hunk 4]
```diff
@@ -78,6 +99,9 @@ func (p *InsertInto) Execute(ctx *sql.Context) (int, error) {
 		}
 
 		if !found {
+			if !f.Nullable {
```
Comment (zachmu) [inline sql/plan/insert.go]: This isn't right -- some non-nullable columns will have a default value that will be set if none is provided in an insert. Integrators will need to catch this on their own, I think. I don't think you can fix this completely on this pass.

File: sql/plan/insert.go [hunk 5]
```diff
@@ -129,7 +160,19 @@ func (p *InsertInto) WithChildren(children ...sql.Node) (sql.Node, error) {
 		return nil, sql.ErrInvalidChildrenNumber.New(p, len(children), 2)
 	}
 
-	return NewInsertInto(children[0], children[1], p.Columns), nil
+	return NewInsertInto(children[0], children[1], p.Columns, p.Ignore), nil
+}
+
+func (p *InsertInto) ValidateValueCount(ctx *sql.Context) error {
```
Comment (zachmu) [inline sql/plan/insert.go]: Probably should not be exported

File: sql/plan/insert.go [hunk 6]
```diff
@@ -11,18 +11,26 @@ import (
...
 // InsertInto is a node describing the insertion into some table.
 type InsertInto struct {
 	BinaryNode
 	Columns []string
+	Ignore  bool
 }
 
 // NewInsertInto creates an InsertInto node.
-func NewInsertInto(dst, src sql.Node, cols []string) *InsertInto {
+func NewInsertInto(dst, src sql.Node, cols []string, ignore bool) *InsertInto {
```
Comment (zachmu) [inline sql/plan/insert.go]: The new param (and existing ones) need documenting.

File: sql/core.go [hunk 7]
```diff
@@ -198,8 +198,8 @@ type IndexableTable interface {
 
 // Inserter allow rows to be inserted in them.
 type Inserter interface {
-	// Insert the given row.
-	Insert(*Context, Row) error
+	// Insert the given row. Returns true if the row did not exist.
+	Insert(*Context, Row) (bool, error)
```
Comment (zachmu) [inline sql/core.go]: I would leave the interface alone for the time being -- makes it much easier to accept this PR back to the upstream.

The semantics of Ignore are actually more general than just "ignore if this row already exists" -- it's really about ignoring any error in a batch of inserts:

https://dev.mysql.com/doc/refman/5.5/en/insert.html

So I think for a complete solution, we shouldn't change this interface, but we should define error types that integrators can return under different circumstances (nullness violation, duplicate keys, etc). This isn't just a simple bug fix, but an interface change. Let's start with a test case that doesn't work currently, and start a discussion with them about the right way to fix it.

Comment (Hydrocharged) [reply sql/core.go]: So as per the conversation held IRL, we'll just skip this altogether for now and come back to it at a later date.

File: engine_test.go [hunk 8]
```diff
@@ -1755,16 +1755,47 @@ func TestOrderByColumns(t *testing.T) {
...
-		[]sql.Row{{int64(1)}},
-	)
+	var insertions = []struct {
+		insertQuery    string
+		expectedInsert []sql.Row
+		selectQuery    string
+		expectedSelect []sql.Row
+	}{
+		{
+			"INSERT INTO mytable (s, i) VALUES ('x', 999);",
```
Comment (zachmu) [inline engine_test.go]: You should have tests of all the supported data types. I found bugs with handling floats in the past.

Also test inserting literal NULL values.

File: sql/plan/insert.go [hunk 9]
```diff
@@ -78,8 +95,7 @@ func (p *InsertInto) Execute(ctx *sql.Context) (int, error) {
 		}
 
 		if !found {
-			def, _ := f.Type.Convert(nil)
-			projExprs[i] = expression.NewLiteral(def, f.Type)
+			projExprs[i] = expression.NewLiteral(nil, f.Type)
```
Comment (zachmu) [inline sql/plan/insert.go]: This is a behavior change. Before it was effectively getting a default value for each type and inserting it. You're changing it to literal nils, which is very different (but more correct).

You need test cases demonstrating that this is the correct behavior.

Comment (Hydrocharged) [reply sql/plan/insert.go]: Before it was getting the _default_, which wasn't quite the same depending on the type. For example, the default for an `int` would be 0, but the default for a `bool` was `nil`. I'm not too sure on how to write a test that this is correct behavior though...

Comment (zachmu) [reply sql/plan/insert.go]: Right, you need to write tests that explicitly insert NULL literals into columns of various types, then verify that these columns contain NULL aftewards. AFAIK they don't have any tests like that currently.

Comment (Hydrocharged) [reply sql/plan/insert.go]: Ohhhh gotchaaa

File: engine_test.go [hunk 10]
```diff
@@ -919,6 +956,23 @@ var queries = []struct {
...
+			{"int16"},
+			{"int32"},
+			{"int64"},
+			{"uint8"},
+			{"uint16"},
+			{"uint32"},
+			{"uint64"},
+			{"float32"},
+			{"float64"},
+			{"timestamp"},
```
Comment (zachmu) [inline engine_test.go]: Some of these are reserved words. Might want to name them something slightly different to make things easier.
